### PR TITLE
[WIP] current progress on GitHub cog

### PR DESCRIPTION
**Pull request by centipeda**
still very much a work in progress

### CODE REVIEW

File: github.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+        self.bot.loop.create_task(self._wait_for_updates())
+
+    def _format_issue(self, issue, repo):
+        """Formats a GitHub issue nicely so it can be printed in Discord. issue is the dict returned by a GitHub
+        API request for a given Issue."""
+        string = "New issue on `{}` by {}: **{}**\n{}"
+        fullstring = string.format(repo, issue["user"]["login"], issue["title"], issue["html_url"])
+        return fullstring
+
+    async def _update_server_list(self):
```
Comment (Chovin) [inline github.py:37]: this doesn't need to be async

also, why are you loading all of the bot's servers into the settings?

File: github.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    """Accesses GitHub."""
+
+    def __init__(self, bot):
+        self.bot = bot
+        self.settingPath = "data/github/settings.json"
+        self.settings = dataIO.load_json(self.settingPath)
+        # async functions can't be called without being in another async function
+        # so _log_in and _wait_for_issue are added to the bot's event loop
+        self.bot.loop.create_task(self._update_server_list())
+        self.bot.loop.create_task(self._log_in())
```
Comment (Chovin) [inline github.py:27]: be careful of requests being made before the cog has logged in.

File: github.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+                # checks if current second count since startup is evenly
+                # divisible by each server's set interval, effectively firing
+                # digests at each server's interval
+                intrval = self.settings["servers"][server.id]["interval"]
+                if referenceCount % intrval == 0:
+                    # notification when scheduled digests fire
+                    print("Scheduled digests firing!")
+                    await self._create_digest(server)
+            await asyncio.sleep(1)
+            referenceCount += 1
```
Comment (Chovin) [inline github.py:131]: this is not an accurate clock. if you're looking for more accuracy, please see how other cogs do it
here's an example in [26's trigger](https://github.com/Twentysix26/26-Cogs/blob/master/trigger/trigger.py#L502)

File: github.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+                        await self.bot.send_message(self._format_issue(issue, repo), dest)
+                elif self.settings["notification_method"][0] == "webhook":
+                    await self._fire_hooks(server)
+
+    async def _wait_for_updates(self):
+        """Checks for updates from assigned GitHub repos at given interval."""
+        await self.bot.wait_until_ready()
+        # loops until bot stops functioning
+        referenceCount = 0
+        while not self.bot.is_closed:
```
Comment (Chovin) [inline github.py:120]: `!reload` will create duplicates of this task.
please see how trigger does it [#L459](https://github.com/Twentysix26/26-Cogs/blob/master/trigger/trigger.py#L459), [#L479](https://github.com/Twentysix26/26-Cogs/blob/master/trigger/trigger.py#L479)
`__unload` is called whenever a cog is unloaded afaik

**Review by Chovin — COMMENTED** [review]
just some comments.
1st time trying this code review thing

File: github.py (hunk 1 above)
Comment (centipeda) [reply github.py:37]: ...Huh. Why _is_ that async?

**Comment by Joshy3282** [comment]
😳😳

**Pull request closed without merging**